README.md
# cloj-fin

A clojure project designed to help teach me how to use clojure as well as up my skills when
it comes to financial and stock analysis.

## Usage

This application should only be thought of as a proof of concept for how clojure can be used as
a tool for financial analysis of the stock market. I am in no way liable for anything that
happens from actually using this information.

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/cloj-fin/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/cloj-fin/compare/0.1.0...0.1.1


test/cloj_fin/core_test.clj
(ns cloj-fin.core-test
  (:require [clojure.test :refer :all]
            [cloj-fin.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/yahoo_fin/history.clj
(ns yahoo-fin.history
  "functionality for pulling data from the history tab of yahoo finance"
  (:import (org.jsoup Jsoup))
  (:require [clj-http.client :as client]))

; NEED TO MAKE SURE THAT I AM USING MULTILINE DOC STRINGS WITHIN FUNCTIONS SP THAT THEY ARE
; DOCUMENTED AS BEST AS POSSIBLE

(defn unix-years-from-today 
  "calculates the unix time stamp date n years away from current day"
  [years] 
  (-> (System/currentTimeMillis) 
      (- (* 31556952000 years)) 
      (quot 1000) 
      (Math/floor) 
      (int)))


(defn req-url 
  "returns the request url given a valid url ticker value"
  [ticker] 
  (format "https://finance.yahoo.com/quote/%s/history" ticker))


(defn req-body
  "returns the request headers and parameters"
  [] 
  {:headers {"User-Agent" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0"
             "Accept" "text/html"
             "Accept-Language" "en-US,en;q=0.5"
             "Pragma" "no-cache"
             "Connection" "keep-alive"
             "Cache-Control" "no-cache"}
   ; for the time range we are only looking a max of 5 years back since anything 
   ; further is probably not needed for our given use case
   :query-params {"period1"(unix-years-from-today 5) 
                  "period2" (unix-years-from-today 0)
                  "interval" "1d"} 
   ; the below two lines are needed since we don't care about cookies with
   ; this specific implementation
   :decode-cookies false 
   :cookie-policy :none})


(defn req-html
  "requests html given a valid ticker symbol and handle when something goes wrong"
  [ticker]
  (-> (client/get (req-url "AAPL") (req-body))))


(defn html->data
  [html]
  (println "we need to parse out the data"))


; MAIN FUNCTION
(defn -main []
  (let [resp (req-html "AAPL")]
    (println (get resp :status))))


src/cloj_fin/core.clj
(ns cloj-fin.core
  (:import (org.jsoup Jsoup))
  (:require [clj-http.client :as client]))

; need to recieve some data from an api endpoint and once we recieve this data then when
; can start to calculate certain statistics off of it such as different types of moving
; averages. It would then be nice if we implement some kind of regression so that thematical
; data can then be used to make assumptions about the movements.
; it looks like we are going to have to use some java APIs to get the http requests working
; in the way that I want it to since I'm not sure how to write an HTTP client from scratch

(defn yahoo-fin-url [ticker tab]
  "creates a url endpoint for yahoo finanance given the ticker and tab"
  (let [tab-end (case tab
                  "stats" "/key-statistics"
                  "history" "/history"
                  "financials" "/financials"
                  "analysis" "/analysis"
                  "")]
    (format "https://finance.yahoo.com/quote/%s%s?p=%s" ticker tab-end ticker)))


; I want to end up doing more error handling here, but I'm not sure the best way
; to have these errors be handled.
(defn get-url [url]
  "gets data from a specific url and handles any errors within the response"
  (let [resp (client/get url)]
    (resp)))

; this should give me the full request that I want to yahoo finance for the historical data
(def h {"User-Agent" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:94.0) Gecko/20100101 Firefox/94.0"
        "Accept" "text/html"
        "Accept-Language" "en-US,en;q=0.5"})

(def qp {"period1" "1606618744"
         "period2" "1638154744"
         "interval" "1d"})

(client/get "https://finance.yahoo.com/quote/AAPL/history" 
            {:headers h :query-params qp})


; parse the different data tabs on yahoo finance
; we will have an individual function for each of the different tabs since
; they will all have different data structures to them
(defn yahoo-fin-parser-summary [html]
  "takes in an html string from the yahoo finance summary tab and parses out the data"
  (let [soup (Jsoup/parse html)
        data (as-> soup s
               (.getElementsByTag s "table")
               (take 2 s)
               (for [table s, 
                     row (.getElementsByTag table "tr"), 
                     data (.getElementsByTag row "td")] 
                 (.text data))
               (apply hash-map s))]
    {:data data}))

(defn yahoo-fin-parser-stats [html]
  "takes in an html string from the yahoo finance stats tab and parses out the data"
  (let [soup (Jsoup/parse html)
        data (as-> soup s)]
    {:data data}))


; pulls everything togther into a single function
(defn yahoo-fin-data [ticker tab]
  "creates the url, pulls the data, and parses the data from yahoo finance"
  (let [url (yahoo-fin-url ticker tab)
        html (get (client/get url) :body)
        parser (case tab
                 "stats" yahoo-fin-parser-stats
                 yahoo-fin-parser-summary)]
    (parser html)))


; I am also going to want to collect information about the company from Yahoo finance or a
; similar website so that we can make more informed decisions on the different stocks
; - continuation trade setup
; - reversal trade setup
; - range-bound trade setup
; - breakout trade setup

; this might a another good website for https://tradingsim.com/blog/day-trading-setups/
; trying to automate looking for different setups


(defn -main []
  (let [data (yahoo-fin-data "AAPL" "summary")]
    (println (-> data (get :data))))
  (let [data (yahoo-fin-data "AAPL" "stats")]
    (println (-> data (get :data))))
  )

